OpenSim/Simulation/Test/testRealizeDynamics.cpp

```diff
@@ -0,0 +1,101 @@
+#include <OpenSim/Simulation/Control/ControlSetController.h>
+#include <OpenSim/Simulation/Model/Model.h>
+#include <OpenSim/Common/LoadOpenSimLibrary.h>
+#include <OpenSim/Auxiliary/auxiliaryTestFunctions.h>
+
+using namespace OpenSim;
+using namespace std;
+
+//==============================================================================
+// testRealizeDynamics tests that the results produced in every stage are 
+// repeatable - this test is designed primarily to make sure that there are no 
+// memory races during realizeDynamics that can cause time-dependent results 
+// with regards to thread timing.
+//==============================================================================
+void testRealizeDynamics(const string& modelFile);
+
+static const int MAX_N_TRIES = 100;
+
+int main()
+{
+    try {
+        LoadOpenSimLibrary("osimActuators");
+        testRealizeDynamics("arm26.osim");
+    }
+    catch (const Exception& e) {
+        cout << "testRealizeDynamics failed: ";
+        e.print(cout);
+        return 1;
+    }
+    catch (const std::exception& e) {
+        cout << "testRealizeDynamics failed: " << e.what() << endl;
+        return 1;
+    }
+    cout << "Done" << endl;
+    return 0;
+}
+
+//==============================================================================
+// Test Cases
+//==============================================================================
+void testRealizeDynamics(const string& modelFile)
+{
+    using namespace SimTK;
+    using namespace std;
+
+    //==========================================================================
+    // Setup OpenSim model
+    Model model(modelFile);
+
+    ControlSetController* controller = new ControlSetController();
+    controller->setControlSetFileName("arm26_StaticOptimization_controls.xml");
+    model.addController(controller);
+
+    //==========================================================================
+    // Test Accuracy of Dynamics
+    
+    //Realize to Stage::Dynamics once to obtain a Dyanmics "answer"
+    State& state = model.initSystem();
+    model.getSystem().realize(state, Stage::Dynamics);
+    const MultibodySystem& mbs = model.getMultibodySystem();
+    Vector_<SpatialVec>&   rigidBodyForcesAnswer = mbs.updRigidBodyForces(state,
+                                                               Stage::Dynamics);
+    
+    string answerString = rigidBodyForcesAnswer.toString(); 
+    //Constantly realize to Stage::Dynamics to make sure that our answers are
+    //consistent
+    for(int x = 0; x < 100; x++)
+    {
+      State& state = model.initSystem();
+
+      model.getSystem().realize(state, Stage::Dynamics);
+      const MultibodySystem& mbs = model.getMultibodySystem();
+      Vector_<SpatialVec>   tempRigidBodyForces = mbs.updRigidBodyForces(state,
+                                                               Stage::Dynamics);
+      ASSERT((int)(answerString == tempRigidBodyForces.toString()) == 1);
```

Comment:
I also think that testing `arm26` is sufficient. When I had a very small data race, the test would fail every time. I think that testing on a more complicated model would only make the test take longer to finish but wouldn't be much of a benefit.